File: services/synthetic_checkout_service.py
from __future__ import annotations

from dataclasses import dataclass
import argparse
import json
import logging
import os
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any, Callable
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen


log = logging.getLogger("synthetic_checkout_service")


class LokiPushError(Exception):
    """Raised when the synthetic service cannot send logs to Loki."""


@dataclass(frozen=True)
class ServiceConfig:
    host: str = "0.0.0.0"
    port: int = 8081
    loki_url: str = "http://localhost:3100"
    service_name: str = "checkout-api"


def build_loki_payload(
    *,
    service_name: str,
    checkout_id: str,
    timestamp_ns: str | None = None,
) -> dict[str, Any]:
    timestamp = timestamp_ns or str(int(time.time() * 1_000_000_000))
    return {
        "streams": [
            {
                "stream": {
                    "service": service_name,
                    "component": "synthetic-checkout",
                    "scenario": "payment-timeout",
                },
                "values": [
                    [timestamp, f"checkout_id={checkout_id} payment timeout after 3000ms"],
                    [timestamp, f"checkout_id={checkout_id} retry queue depth increasing for payment gateway"],
                    [timestamp, f"checkout_id={checkout_id} customer checkout failed after payment retries"],
                ],
            }
        ]
    }


def push_loki_logs(
    loki_url: str,
    payload: dict[str, Any],
    *,
    opener: Callable[..., Any] = urlopen,
    timeout_seconds: int = 10,
) -> None:
    body = json.dumps(payload).encode("utf-8")
    request = Request(
        f"{loki_url.rstrip('/')}/loki/api/v1/push",
        data=body,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with opener(request, timeout=timeout_seconds) as response:
            response.read()
    except HTTPError as error:
        raise LokiPushError(f"Loki HTTP error: {error.code}.") from error
    except URLError as error:
        raise LokiPushError(f"Loki request failed: {error.reason}.") from error
    except TimeoutError as error:
        raise LokiPushError("Loki request timed out.") from error


def serve(config: ServiceConfig) -> None:
    handler = _handler_for(config)
    httpd = ThreadingHTTPServer((config.host, config.port), handler)
    log.info("Starting synthetic checkout service on %s:%s.", config.host, config.port)
    httpd.serve_forever()


def _handler_for(config: ServiceConfig):
    class SyntheticCheckoutHandler(BaseHTTPRequestHandler):
        def do_GET(self) -> None:
            if self.path != "/health":
                self._write_json(404, {"status": "error", "error": "not_found"})
                return
            self._write_json(200, {"status": "ok", "service": config.service_name})

        def do_POST(self) -> None:
            if self.path != "/checkout":
                self._write_json(404, {"status": "error", "error": "not_found"})
                return

            try:
                request_payload = self._read_request_payload()
            except ValueError as error:
                self._write_json(400, {"status": "error", "error": str(error)})
                return

            checkout_id = str(request_payload.get("checkout_id") or f"checkout-{int(time.time())}")
            payload = build_loki_payload(service_name=config.service_name, checkout_id=checkout_id)
            try:
                push_loki_logs(config.loki_url, payload)
            except LokiPushError:
                log.exception("Failed to push synthetic checkout logs to Loki.")
                self._write_json(502, {"status": "error", "error": "loki_push_failed"})
                return

            self._write_json(
                202,
                {
                    "status": "accepted",
                    "service": config.service_name,
                    "checkout_id": checkout_id,
                    "log_count": len(payload["streams"][0]["values"]),
                },
            )

        def log_message(self, _format: str, *_args: Any) -> None:
            return

        def _read_request_payload(self) -> dict[str, Any]:
            length = int(self.headers.get("Content-Length", "0"))
            if length > 100_000:
                raise ValueError("payload_too_large")
            if length == 0:
                return {}
            try:
                payload = json.loads(self.rfile.read(length).decode("utf-8"))
            except (UnicodeDecodeError, json.JSONDecodeError) as error:
                raise ValueError("invalid_json") from error
            if not isinstance(payload, dict):
                raise ValueError("payload_must_be_object")
            return payload

        def _write_json(self, status: int, payload: dict[str, Any]) -> None:
            body = json.dumps(payload).encode("utf-8")
            self.send_response(status)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)

    return SyntheticCheckoutHandler


def config_from_env() -> ServiceConfig:
    return ServiceConfig(
        host=os.environ.get("HOST", "0.0.0.0"),
        port=int(os.environ.get("PORT", "8081")),
        loki_url=os.environ.get("LOKI_URL", "http://localhost:3100"),
        service_name=os.environ.get("SERVICE_NAME", "checkout-api"),
    )


def main() -> int:
    parser = argparse.ArgumentParser(description="Run the synthetic checkout service.")
    parser.add_argument("--host", default=None)
    parser.add_argument("--port", type=int, default=None)
    parser.add_argument("--loki-url", default=None)
    parser.add_argument("--service-name", default=None)
    args = parser.parse_args()

    logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(name)s %(message)s")
    env_config = config_from_env()
    config = ServiceConfig(
        host=args.host or env_config.host,
        port=args.port or env_config.port,
        loki_url=args.loki_url or env_config.loki_url,
        service_name=args.service_name or env_config.service_name,
    )
    serve(config)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: services/synthetic-checkout-service.ts
import { createServer, type IncomingMessage, type Server, type ServerResponse } from "node:http";
import pino from "pino";

const logger = pino({ name: "synthetic-checkout-service", level: process.env.LOG_LEVEL ?? "info" });

export class LokiPushError extends Error {
  constructor(message: string, options?: ErrorOptions) {
    super(message, options);
    this.name = "LokiPushError";
  }
}

export interface ServiceConfig {
  host: string;
  port: number;
  lokiUrl: string;
  serviceName: string;
}

export interface LokiPayload {
  streams: Array<{
    stream: Record<string, string>;
    values: Array<[string, string]>;
  }>;
}

export interface SyntheticService {
  server: Server;
  ready: Promise<void>;
  closed: Promise<void>;
  close(): Promise<void>;
}

type ScenarioName = "payment-timeout" | "capacity-saturation" | "bad-deploy-latency";

const scenarioLogLines: Record<ScenarioName, string[]> = {
  "payment-timeout": [
    "checkout_id={incident_id} payment timeout after 3000ms",
    "checkout_id={incident_id} retry queue depth increasing for payment gateway",
    "checkout_id={incident_id} customer checkout failed after payment retries",
  ],
  "capacity-saturation": [
    "incident_id={incident_id} search worker cpu=94 queue_depth=438",
    "incident_id={incident_id} search-api p95 latency elevated while workers saturated",
    "incident_id={incident_id} autoscaling backlog still above baseline",
  ],
  "bad-deploy-latency": [
    "incident_id={incident_id} checkout-api p95 latency elevated after v2.19.0 traffic ramp",
    "incident_id={incident_id} checkout retry attempts increased for checkout-api requests",
    "incident_id={incident_id} checkout error rate above baseline for new version",
  ],
};

const routes: Record<string, {
  scenario: ScenarioName;
  service: (config: ServiceConfig) => string;
  idField: string;
  idPrefix: string;
}> = {
  "/checkout": {
    scenario: "payment-timeout",
    service: (config) => config.serviceName,
    idField: "checkout_id",
    idPrefix: "checkout",
  },
  "/capacity": {
    scenario: "capacity-saturation",
    service: () => "search-api",
    idField: "incident_id",
    idPrefix: "capacity",
  },
  "/bad-deploy": {
    scenario: "bad-deploy-latency",
    service: () => "checkout-api",
    idField: "incident_id",
    idPrefix: "bad-deploy",
  },
};

export function buildLokiPayload(input: {
  serviceName: string;
  checkoutId: string;
  timestampNs?: string;
  scenario?: ScenarioName;
}): LokiPayload {
  const scenario = input.scenario ?? "payment-timeout";
  const lines = scenarioLogLines[scenario];
  const timestamp = input.timestampNs ?? String(BigInt(Date.now()) * 1_000_000n);

  return {
    streams: [{
      stream: {
        service: input.serviceName,
        component: "synthetic-checkout",
        scenario,
      },
      values: lines.map((line) => [
        timestamp,
        line.replaceAll("{incident_id}", input.checkoutId),
      ]),
    }],
  };
}

export async function pushLokiLogs(
  lokiUrl: string,
  payload: LokiPayload,
  options: {
    timeoutSeconds?: number;
    fetcher?: typeof fetch;
  } = {},
): Promise<void> {
  const timeoutSeconds = options.timeoutSeconds ?? 10;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutSeconds * 1000);

  try {
    const fetcher = options.fetcher ?? fetch;
    const response = await fetcher(`${lokiUrl.replace(/\/$/, "")}/loki/api/v1/push`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
      signal: controller.signal,
    });
    if (!response.ok) {
      throw new LokiPushError(`Loki HTTP error: ${response.status}.`);
    }
    await response.arrayBuffer();
  } catch (error) {
    if (error instanceof LokiPushError) {
      throw error;
    }
    throw new LokiPushError(`Loki request failed: ${error instanceof Error ? error.message : String(error)}.`, {
      cause: error,
    });
  } finally {
    clearTimeout(timer);
  }
}

export function startSyntheticCheckoutService(config: ServiceConfig): SyntheticService {
  const server = createServer(async (request, response) => {
    try {
      await routeRequest(request, response, config);
    } catch (error) {
      logger.error({ error: error instanceof Error ? error.message : String(error) }, "Unhandled synthetic service request error");
      writeJson(response, 500, { status: "error", error: "internal_server_error" });
    }
  });

  const ready = new Promise<void>((resolve, reject) => {
    server.once("error", reject);
    server.listen(config.port, config.host, () => {
      server.off("error", reject);
      logger.info({ host: config.host, port: config.port }, "Synthetic checkout service ready");
      resolve();
    });
  });
  const closed = new Promise<void>((resolve) => server.once("close", resolve));

  return {
    server,
    ready,
    closed,
    close: async () => closeServer(server),
  };
}

export function configFromEnv(env: NodeJS.ProcessEnv = process.env): ServiceConfig {
  return {
    host: env.HOST ?? "0.0.0.0",
    port: Number.parseInt(env.PORT ?? "8081", 10),
    lokiUrl: env.LOKI_URL ?? "http://localhost:3100",
    serviceName: env.SERVICE_NAME ?? "checkout-api",
  };
}

async function routeRequest(
  request: IncomingMessage,
  response: ServerResponse,
  config: ServiceConfig,
): Promise<void> {
  if (request.method === "GET" && request.url === "/health") {
    writeJson(response, 200, { status: "ok", service: config.serviceName });
    return;
  }

  const route = request.url ? routes[request.url] : undefined;
  if (request.method !== "POST" || route === undefined) {
    writeJson(response, 404, { status: "error", error: "not_found" });
    return;
  }

  let requestPayload: Record<string, unknown>;
  try {
    requestPayload = await readJsonObject(request);
  } catch (error) {
    writeJson(response, 400, { status: "error", error: error instanceof Error ? error.message : String(error) });
    return;
  }

  const incidentId = String(requestPayload[route.idField] ?? `${route.idPrefix}-${Math.floor(Date.now() / 1000)}`);
  const payload = buildLokiPayload({
    serviceName: route.service(config),
    checkoutId: incidentId,
    scenario: route.scenario,
  });

  try {
    await pushLokiLogs(config.lokiUrl, payload);
  } catch (error) {
    logger.error({ error: error instanceof Error ? error.message : String(error) }, "Failed to push synthetic logs to Loki");
    writeJson(response, 502, { status: "error", error: "loki_push_failed" });
    return;
  }

  writeJson(response, 202, {
    status: "accepted",
    service: route.service(config),
    scenario: route.scenario,
    [route.idField]: incidentId,
    log_count: payload.streams[0]?.values.length ?? 0,
  });
}

function readJsonObject(request: IncomingMessage): Promise<Record<string, unknown>> {
  return new Promise((resolve, reject) => {
    const length = Number.parseInt(request.headers["content-length"] ?? "0", 10);
    if (length > 100_000) {
      reject(new Error("payload_too_large"));
      request.destroy();
      return;
    }
    if (length === 0) {
      resolve({});
      return;
    }

    let body = "";
    request.setEncoding("utf8");
    request.on("data", (chunk: string) => {
      body += chunk;
    });
    request.on("error", reject);
    request.on("end", () => {
      try {
        const payload: unknown = JSON.parse(body);
        if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
          reject(new Error("payload_must_be_object"));
          return;
        }
        resolve(payload as Record<string, unknown>);
      } catch {
        reject(new Error("invalid_json"));
      }
    });
  });
}

function writeJson(response: ServerResponse, status: number, payload: Record<string, unknown>): void {
  const body = JSON.stringify(payload);
  response.writeHead(status, {
    "Content-Type": "application/json",
    "Content-Length": Buffer.byteLength(body),
  });
  response.end(body);
}

function closeServer(server: Server): Promise<void> {
  if (!server.listening) {
    return Promise.resolve();
  }
  return new Promise((resolve, reject) => {
    server.close((error) => {
      if (error) {
        reject(error);
        return;
      }
      resolve();
    });
  });
}

async function main(): Promise<number> {
  const service = startSyntheticCheckoutService(configFromEnv());
  await service.ready;
  await service.closed;
  return 0;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().then(
    (code) => {
      process.exitCode = code;
    },
    (error: unknown) => {
      logger.error({ error: error instanceof Error ? error.message : String(error) }, "Synthetic service failed");
      process.exitCode = 1;
    },
  );
}
